File: email-service/src/types/storage.ts
/**
 * Storage Type Definitions
 * 
 * This module defines TypeScript interfaces for S3-compatible storage operations,
 * bucket configurations, and document metadata used by the Email Service.
 * It provides type definitions for securely storing email attachments in the
 * document repository with AES-256 encryption.
 *
 * The Email Service uses these types to store attachments in S3-compatible storage
 * as specified in the technical requirements. Documents are stored with appropriate
 * metadata and the storage system implements proper error handling and retry logic.
 */

import { IResult, IAsyncResult } from './common';

/**
 * S3 storage configuration
 */
export interface IS3Config {
  /** S3 endpoint URL */
  endpoint: string;
  
  /** S3 region */
  region: string;
  
  /** Default bucket name */
  defaultBucket: string;
  
  /** Access key ID */
  accessKeyId: string;
  
  /** Secret access key */
  secretAccessKey: string;
  
  /** Whether to force path style addressing */
  forcePathStyle?: boolean;
  
  /** Maximum number of concurrent connections */
  maxConnections?: number;
  
  /** Connection timeout in milliseconds */
  connectionTimeout?: number;
  
  /** Whether to use SSL */
  sslEnabled?: boolean;
  
  /** Whether to accelerate endpoint */
  accelerateEndpoint?: boolean;
}

/**
 * Storage options for S3 operations
 */
export interface IStorageOptions {
  /** 
   * Server-side encryption method
   * AES256 is required as per section 0.2.5
   */
  encryption?: 'AES256' | 'aws:kms';
  
  /** KMS key ID if using aws:kms encryption */
  kmsKeyId?: string;
  
  /** Content type of the object */
  contentType?: string;
  
  /** Content disposition */
  contentDisposition?: string;
  
  /** Cache control directives */
  cacheControl?: string;
  
  /** Custom metadata */
  metadata?: Record<string, string>;
  
  /** Object tags */
  tags?: Record<string, string>;
  
  /** Storage class */
  storageClass?: 'STANDARD' | 'REDUCED_REDUNDANCY' | 'STANDARD_IA' | 'ONEZONE_IA' | 'INTELLIGENT_TIERING' | 'GLACIER' | 'DEEP_ARCHIVE';
  
  /** ACL for the object */
  acl?: 'private' | 'public-read' | 'public-read-write' | 'authenticated-read' | 'aws-exec-read' | 'bucket-owner-read' | 'bucket-owner-full-control';
}

/**
 * Document metadata for S3 storage
 */
export interface IDocumentMetadata {
  /** Unique identifier for the document */
  documentId: string;
  
  /** Original filename */
  filename: string;
  
  /** File MIME type */
  mimeType: string;
  
  /** File size in bytes */
  fileSize: number;
  
  /** Document type classification */
  documentType?: string;
  
  /** Classification confidence score (0-1) */
  confidenceScore?: number;
  
  /** Email message ID that contained this document */
  emailMessageId: string;
  
  /** Email sender information */
  sender: {
    /** Sender email address */
    email: string;
    
    /** Sender name if available */
    name?: string;
  };
  
  /** Email subject */
  subject: string;
  
  /** Email received timestamp */
  receivedAt: string;
  
  /** Upload timestamp */
  uploadedAt: string;
  
  /** Processing status */
  status: 'new' | 'processing' | 'processed' | 'failed';
  
  /** Checksum for file integrity verification */
  checksum: string;
  
  /** Checksum algorithm used */
  checksumAlgorithm: 'md5' | 'sha1' | 'sha256';
  
  /** Application ID if this document is associated with an application */
  applicationId?: string;
}

/**
 * Bucket configuration for different environments
 * As specified in section 0.2.5
 */
export interface IBucketConfig {
  /** Bucket name */
  name: string;
  
  /** Whether versioning is enabled */
  versioning: boolean;
  
  /** Default encryption method */
  encryption: 'AES256' | 'aws:kms' | 'none';
  
  /** KMS key ID if using aws:kms encryption */
  kmsKeyId?: string;
  
  /** Lifecycle rules */
  lifecycleRules?: ILifecycleRule[];
  
  /** Bucket policy */
  policy?: Record<string, unknown>;
  
  /** CORS configuration */
  cors?: ICorsRule[];
}

/**
 * Lifecycle rule for S3 buckets
 */
export interface ILifecycleRule {
  /** Rule ID */
  id: string;
  
  /** Whether the rule is enabled */
  enabled: boolean;
  
  /** Prefix filter */
  prefix?: string;
  
  /** Tag filters */
  tags?: Record<string, string>;
  
  /** Expiration configuration */
  expiration?: {
    /** Days after which objects expire */
    days?: number;
    
    /** Date after which objects expire */
    date?: string;
  };
  
  /** Transition configuration */
  transitions?: Array<{
    /** Days after which objects transition */
    days: number;
    
    /** Storage class to transition to */
    storageClass: string;
  }>;
}

/**
 * CORS rule for S3 buckets
 */
export interface ICorsRule {
  /** Allowed origins */
  allowedOrigins: string[];
  
  /** Allowed methods */
  allowedMethods: Array<'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD'>;
  
  /** Allowed headers */
  allowedHeaders: string[];
  
  /** Exposed headers */
  exposedHeaders?: string[];
  
  /** Max age in seconds */
  maxAgeSeconds?: number;
}

/**
 * Signed URL options
 */
export interface ISignedUrlOptions {
  /** URL expiration time in seconds */
  expiresIn: number;
  
  /** HTTP method */
  method: 'GET' | 'PUT' | 'POST' | 'DELETE';
  
  /** Content type (for PUT/POST) */
  contentType?: string;
  
  /** Additional query parameters */
  queryParams?: Record<string, string>;
}

/**
 * Storage operation result
 */
export interface IStorageOperationResult {
  /** Success status */
  success: boolean;
  
  /** S3 key of the stored object */
  key?: string;
  
  /** S3 bucket where the object is stored */
  bucket?: string;
  
  /** ETag of the stored object */
  etag?: string;
  
  /** Version ID if versioning is enabled */
  versionId?: string;
  
  /** Error message if operation failed */
  error?: string;
  
  /** Detailed error information */
  errorDetails?: Record<string, unknown>;
}

/**
 * Storage service interface
 */
export interface IStorageService {
  /**
   * Upload a file to S3 storage
   * 
   * @param filePath - Local file path to upload
   * @param key - S3 object key
   * @param metadata - Document metadata
   * @param options - Storage options
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  uploadFile(
    filePath: string,
    key: string,
    metadata: IDocumentMetadata,
    options?: IStorageOptions,
    bucket?: string
  ): IAsyncResult<IStorageOperationResult>;
  
  /**
   * Upload a buffer to S3 storage
   * 
   * @param buffer - Buffer to upload
   * @param key - S3 object key
   * @param metadata - Document metadata
   * @param options - Storage options
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  uploadBuffer(
    buffer: Buffer,
    key: string,
    metadata: IDocumentMetadata,
    options?: IStorageOptions,
    bucket?: string
  ): IAsyncResult<IStorageOperationResult>;
  
  /**
   * Download a file from S3 storage
   * 
   * @param key - S3 object key
   * @param destinationPath - Local destination path
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  downloadFile(
    key: string,
    destinationPath: string,
    bucket?: string
  ): IAsyncResult<IStorageOperationResult>;
  
  /**
   * Download a file as a buffer from S3 storage
   * 
   * @param key - S3 object key
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  downloadBuffer(
    key: string,
    bucket?: string
  ): IAsyncResult<Buffer>;
  
  /**
   * Delete a file from S3 storage
   * 
   * @param key - S3 object key
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  deleteFile(
    key: string,
    bucket?: string
  ): IAsyncResult<IStorageOperationResult>;
  
  /**
   * Check if a file exists in S3 storage
   * 
   * @param key - S3 object key
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  fileExists(
    key: string,
    bucket?: string
  ): IAsyncResult<boolean>;
  
  /**
   * Get metadata for a file in S3 storage
   * 
   * @param key - S3 object key
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  getMetadata(
    key: string,
    bucket?: string
  ): IAsyncResult<IDocumentMetadata>;
  
  /**
   * Update metadata for a file in S3 storage
   * 
   * @param key - S3 object key
   * @param metadata - Document metadata
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  updateMetadata(
    key: string,
    metadata: Partial<IDocumentMetadata>,
    bucket?: string
  ): IAsyncResult<IStorageOperationResult>;
  
  /**
   * Generate a signed URL for a file in S3 storage
   * 
   * @param key - S3 object key
   * @param options - Signed URL options
   * @param bucket - Optional bucket name (uses default if not specified)
   */
  getSignedUrl(
    key: string,
    options: ISignedUrlOptions,
    bucket?: string
  ): IAsyncResult<string>;
}

/**
 * Default bucket configurations for different environments
 * As specified in section 0.2.5
 */
export const DEFAULT_BUCKET_CONFIGS: Record<string, IBucketConfig> = {
  production: {
    name: 'mca-documents-production',
    versioning: true,
    encryption: 'AES256',
    lifecycleRules: [
      {
        id: 'archive-rule',
        enabled: true,
        transitions: [
          {
            days: 90,
            storageClass: 'STANDARD_IA'
          },
          {
            days: 365,
            storageClass: 'GLACIER'
          }
        ]
      }
    ]
  },
  staging: {
    name: 'mca-documents-staging',
    versioning: true,
    encryption: 'AES256',
    lifecycleRules: [
      {
        id: 'archive-rule',
        enabled: true,
        transitions: [
          {
            days: 30,
            storageClass: 'STANDARD_IA'
          }
        ]
      }
    ]
  },
  development: {
    name: 'mca-documents-development',
    versioning: true,
    encryption: 'AES256',
    lifecycleRules: []
  }
};

/**
 * Default storage options with AES-256 encryption
 * As required in section 0.2.5
 */
export const DEFAULT_STORAGE_OPTIONS: IStorageOptions = {
  encryption: 'AES256',
  acl: 'private',
  storageClass: 'STANDARD',
  cacheControl: 'private, max-age=0, no-cache, no-store'
};

/**
 * Default signed URL expiration time (15 minutes)
 */
export const DEFAULT_SIGNED_URL_EXPIRATION = 15 * 60; // 15 minutes in seconds

File: ocr-service/src/types/storage.py
"""Type definitions for S3-compatible storage operations used by the OCR Service.

This module provides type hints for S3 client configuration, storage operations,
bucket configurations, and document metadata. These types ensure type safety
for all storage operations in the OCR Service.

The OCR Service uses S3-compatible storage with AES-256 encryption to securely
store and retrieve documents for processing.
"""

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Dict, List, Optional, Union, TypeVar, Generic, Any, Callable
from datetime import datetime, timedelta
import os
from pathlib import Path

# Bucket configurations for different environments
BUCKET_CONFIGS = {
    'development': {
        'name': 'mca-documents-development',
        'region': 'us-east-1',
        'versioning_enabled': True,
        'lifecycle_rules': [
            {
                'ID': 'Delete old versions',
                'Status': 'Enabled',
                'NoncurrentVersionExpiration': {'NoncurrentDays': 90}
            }
        ],
        'cors_rules': [
            {
                'AllowedHeaders': ['*'],
                'AllowedMethods': ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
                'AllowedOrigins': ['*'],
                'ExposeHeaders': ['ETag'],
                'MaxAgeSeconds': 3000
            }
        ],
        'encryption': 'AES256'
    },
    'staging': {
        'name': 'mca-documents-staging',
        'region': 'us-east-1',
        'versioning_enabled': True,
        'lifecycle_rules': [
            {
                'ID': 'Delete old versions',
                'Status': 'Enabled',
                'NoncurrentVersionExpiration': {'NoncurrentDays': 90}
            }
        ],
        'cors_rules': [
            {
                'AllowedHeaders': ['*'],
                'AllowedMethods': ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
                'AllowedOrigins': ['*'],
                'ExposeHeaders': ['ETag'],
                'MaxAgeSeconds': 3000
            }
        ],
        'encryption': 'AES256'
    },
    'production': {
        'name': 'mca-documents-production',
        'region': 'us-east-1',
        'versioning_enabled': True,
        'lifecycle_rules': [
            {
                'ID': 'Archive old versions',
                'Status': 'Enabled',
                'NoncurrentVersionTransition': {
                    'NoncurrentDays': 30,
                    'StorageClass': 'STANDARD_IA'
                },
                'NoncurrentVersionExpiration': {'NoncurrentDays': 2555}  # 7 years retention
            }
        ],
        'cors_rules': [
            {
                'AllowedHeaders': ['*'],
                'AllowedMethods': ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
                'AllowedOrigins': ['*'],
                'ExposeHeaders': ['ETag'],
                'MaxAgeSeconds': 3000
            }
        ],
        'encryption': 'AES256'
    }
}


class StorageErrorCode(Enum):
    """Error codes for storage operations."""
    CONNECTION_ERROR = auto()
    AUTHENTICATION_ERROR = auto()
    PERMISSION_DENIED = auto()
    RESOURCE_NOT_FOUND = auto()
    BUCKET_NOT_FOUND = auto()
    OBJECT_NOT_FOUND = auto()
    INVALID_REQUEST = auto()
    TIMEOUT = auto()
    INTERNAL_ERROR = auto()
    UNKNOWN_ERROR = auto()


class EncryptionType(Enum):
    """Encryption types for S3 storage."""
    NONE = "NONE"
    AES256 = "AES256"  # Server-side encryption with Amazon S3-managed keys
    KMS = "aws:kms"    # Server-side encryption with AWS KMS-managed keys


class StorageClass(Enum):
    """Storage classes for S3 objects."""
    STANDARD = "STANDARD"
    REDUCED_REDUNDANCY = "REDUCED_REDUNDANCY"
    STANDARD_IA = "STANDARD_IA"  # Infrequent Access
    ONEZONE_IA = "ONEZONE_IA"    # One Zone Infrequent Access
    INTELLIGENT_TIERING = "INTELLIGENT_TIERING"
    GLACIER = "GLACIER"
    DEEP_ARCHIVE = "DEEP_ARCHIVE"


@dataclass
class S3Credentials:
    """Credentials for S3 authentication.
    
    Attributes:
        access_key: AWS access key ID
        secret_key: AWS secret access key
        session_token: AWS session token (optional)
    """
    access_key: str
    secret_key: str
    session_token: Optional[str] = None


@dataclass
class S3ClientConfig:
    """Configuration for S3 client.
    
    Attributes:
        endpoint_url: S3 endpoint URL
        region: AWS region
        credentials: S3 credentials
        verify_ssl: Whether to verify SSL certificates
        use_path_style: Whether to use path-style addressing
        max_pool_connections: Maximum number of connections in the connection pool
        timeout: Connection timeout in seconds
        retries: Maximum number of retries for failed operations
    """
    endpoint_url: str
    region: str
    credentials: S3Credentials
    verify_ssl: bool = True
    use_path_style: bool = False
    max_pool_connections: int = 10
    timeout: int = 60
    retries: int = 3


@dataclass
class StorageOptions:
    """Options for storage operations.
    
    Attributes:
        encryption: Encryption type to use
        storage_class: Storage class for the object
        metadata: Custom metadata to attach to the object
        content_type: Content type of the object
        content_disposition: Content disposition of the object
        cache_control: Cache control directives for the object
        tags: Tags to attach to the object
        acl: Access control list for the object
    """
    encryption: EncryptionType = EncryptionType.AES256
    storage_class: StorageClass = StorageClass.STANDARD
    metadata: Dict[str, str] = field(default_factory=dict)
    content_type: Optional[str] = None
    content_disposition: Optional[str] = None
    cache_control: Optional[str] = None
    tags: Dict[str, str] = field(default_factory=dict)
    acl: Optional[str] = None


@dataclass
class SignedUrlOptions:
    """Options for generating signed URLs.
    
    Attributes:
        expiration: Expiration time in seconds
        http_method: HTTP method for the URL
        content_type: Content type for PUT/POST requests
        content_disposition: Content disposition for PUT/POST requests
        response_headers: Response headers to override
    """
    expiration: int = 3600  # 1 hour default
    http_method: str = 'GET'
    content_type: Optional[str] = None
    content_disposition: Optional[str] = None
    response_headers: Dict[str, str] = field(default_factory=dict)


@dataclass
class BucketConfig:
    """Configuration for S3 bucket.
    
    Attributes:
        name: Bucket name
        region: AWS region for the bucket
        versioning_enabled: Whether versioning is enabled
        lifecycle_rules: Lifecycle rules for the bucket
        cors_rules: CORS rules for the bucket
        encryption: Default encryption for the bucket
    """
    name: str
    region: str
    versioning_enabled: bool = True
    lifecycle_rules: List[Dict[str, Any]] = field(default_factory=list)
    cors_rules: List[Dict[str, Any]] = field(default_factory=list)
    encryption: EncryptionType = EncryptionType.AES256


@dataclass
class StorageMetadata:
    """Metadata for stored documents.
    
    Attributes:
        document_id: Unique identifier for the document
        application_id: ID of the application this document belongs to
        document_type: Type of the document
        content_type: MIME type of the document
        original_filename: Original filename of the document
        size_bytes: Size of the document in bytes
        md5_hash: MD5 hash of the document content
        created_at: Timestamp when the document was created
        updated_at: Timestamp when the document was last updated
        version_id: S3 version ID if versioning is enabled
        extraction_status: Status of OCR extraction
        extraction_confidence: Overall confidence score of extraction
        tags: Custom tags for the document
        custom_metadata: Additional custom metadata
    """
    document_id: str
    application_id: str
    document_type: str
    content_type: str
    original_filename: str
    size_bytes: int
    md5_hash: str
    created_at: datetime
    updated_at: datetime
    version_id: Optional[str] = None
    extraction_status: Optional[str] = None
    extraction_confidence: Optional[float] = None
    tags: Dict[str, str] = field(default_factory=dict)
    custom_metadata: Dict[str, str] = field(default_factory=dict)
    
    def to_s3_metadata(self) -> Dict[str, str]:
        """Convert to S3 metadata format.
        
        Returns:
            Dict[str, str]: Metadata in S3 format
        """
        metadata = {
            'document-id': self.document_id,
            'application-id': self.application_id,
            'document-type': self.document_type,
            'original-filename': self.original_filename,
            'size-bytes': str(self.size_bytes),
            'md5-hash': self.md5_hash,
            'created-at': self.created_at.isoformat(),
            'updated-at': self.updated_at.isoformat(),
        }
        
        if self.version_id:
            metadata['version-id'] = self.version_id
            
        if self.extraction_status:
            metadata['extraction-status'] = self.extraction_status
            
        if self.extraction_confidence is not None:
            metadata['extraction-confidence'] = str(self.extraction_confidence)
            
        # Add custom metadata
        for key, value in self.custom_metadata.items():
            metadata[f'custom-{key}'] = value
            
        return metadata


T = TypeVar('T')  # Generic type for StorageResult


@dataclass
class StorageResult(Generic[T]):
    """Result of a storage operation.
    
    Attributes:
        success: Whether the operation was successful
        data: Result data if successful
        error_code: Error code if unsuccessful
        error_message: Error message if unsuccessful
        retry_count: Number of retries performed
        timestamp: Timestamp of the operation
    """
    success: bool
    data: Optional[T] = None
    error_code: Optional[StorageErrorCode] = None
    error_message: Optional[str] = None
    retry_count: int = 0
    timestamp: datetime = field(default_factory=datetime.now)
    
    @classmethod
    def success_result(cls, data: T) -> 'StorageResult[T]':
        """Create a successful result.
        
        Args:
            data: Result data
            
        Returns:
            StorageResult[T]: Successful result
        """
        return cls(success=True, data=data)
    
    @classmethod
    def error_result(cls, 
                    error_code: StorageErrorCode, 
                    error_message: str, 
                    retry_count: int = 0) -> 'StorageResult[T]':
        """Create an error result.
        
        Args:
            error_code: Error code
            error_message: Error message
            retry_count: Number of retries performed
            
        Returns:
            StorageResult[T]: Error result
        """
        return cls(
            success=False, 
            error_code=error_code, 
            error_message=error_message,
            retry_count=retry_count
        )


class StorageKey:
    """Utility class for generating and parsing storage keys."""
    
    @staticmethod
    def generate_key(application_id: str, document_id: str, document_type: str, 
                    filename: str) -> str:
        """Generate a storage key for a document.
        
        Args:
            application_id: Application ID
            document_id: Document ID
            document_type: Document type
            filename: Original filename
            
        Returns:
            str: Storage key
        """
        # Sanitize filename to remove problematic characters
        safe_filename = Path(filename).name
        
        # Format: applications/{application_id}/documents/{document_id}/{document_type}/{filename}
        return f"applications/{application_id}/documents/{document_id}/{document_type}/{safe_filename}"
    
    @staticmethod
    def parse_key(key: str) -> Dict[str, str]:
        """Parse a storage key into its components.
        
        Args:
            key: Storage key
            
        Returns:
            Dict[str, str]: Key components
        """
        parts = key.split('/')
        if len(parts) < 5 or parts[0] != 'applications' or parts[2] != 'documents':
            raise ValueError(f"Invalid storage key format: {key}")
            
        return {
            'application_id': parts[1],
            'document_id': parts[3],
            'document_type': parts[4],
            'filename': '/'.join(parts[5:]) if len(parts) > 5 else ''
        }
    
    @staticmethod
    def get_prefix(application_id: str, document_id: Optional[str] = None, 
                  document_type: Optional[str] = None) -> str:
        """Generate a prefix for listing objects.
        
        Args:
            application_id: Application ID
            document_id: Document ID (optional)
            document_type: Document type (optional)
            
        Returns:
            str: Prefix for listing objects
        """
        prefix = f"applications/{application_id}/"
        
        if document_id:
            prefix += f"documents/{document_id}/"
            
            if document_type:
                prefix += f"{document_type}/"
                
        return prefix